src/Compiler/Checking/CheckDeclarations.fs

```diff
@@ -500,18 +500,26 @@ module TcRecdUnionAndEnumDeclarations =
             if not (String.isLeadingIdentifierCharacterUpperCase name) && name <> opNameCons && name <> opNameNil then
                 errorR(NotUpperCaseConstructor(id.idRange))
 
-    let ValidateFieldNames (synFields: SynField list, tastFields: RecdField list) = 
+    let private CheckUnionDuplicateFields (elems: Ident list) =
+        elems |> List.iteri (fun i (uc1: Ident) -> 
+            elems |> List.iteri (fun j (uc2: Ident) -> 
+                if j > i && uc1.idText = uc2.idText then 
```

Comment:
How much elements on average do we expect here?

Reply:
I believe named fields are not really that common(even in the compiler codebase) but an informed guess would be around a dozen on average. This specific helper function is duplicated in multiple places to calculate ident duplication. If performance is a concern we could extract in a follow up PR all of these repeated helper functions and use maybe a better implementation.